proposals/NNNN-method-keypaths.md

```diff
@@ -0,0 +1,643 @@
+[as long ago as 2016](https://forums.swift.org/t/proposal-automating-partial-application-via-wildcards/1284)
+and [as recently as April 2020](https://forums.swift.org/t/thoughts-regarding-the-potential-assignment-of-functions-to-labelled-identifiers/35471).
+It even made an appearance in the venerable
+[SE-0002](https://github.com/apple/swift-evolution/blob/master/proposals/0002-remove-currying.md#alternatives-considered).
+However, it quickly becomes entangled in
+[doubts about Swift’s method reference syntax](https://forums.swift.org/t/require-parameter-names-when-referencing-to-functions/27048)
+that raise thorny source compatibility questions.
+
+Whatever approach such a feature ultimately takes, the **syntax should be uniform across method references and key
+paths**. And whatever syntax compatibility challenges a solution raises for key paths already exist for method
+references. Thus while this document’s proposal to bringing key paths into alignment with method references does not
+_solve_ the problem, neither does it _change_ it.
+
+Given that, we think it best not to let this relatively simple proposal die on the rocks by opening the “partial method
+application” can of worms.
+
+### Mutating methods
+
+This proposal does not support mutating methods, because it is not clear how a caller should access the result of the
+mutation. It seems natural that this code should work:
+
+```swift
+var primes = [2, 3, 5]
+primes[keyPath: \.append](7)
+```
+
+…but what, then, should this do?
+
+```swift
+var primes = [2, 3, 5]
+let appender = primes[keyPath: \.append]
+appender(7)  // Is primes now mutated? Does appender keep a hidden pointer to
+             // a stack-allocated value? What if appender escapes local scope?
+```
+
+Key paths for mutating methods might induce an `inout` param for self:
+
+```swift
+let appender: (inout Array<Int>, Int) -> Void = \Array.append
+appender(primes, 7)
+```
+
+…but that idea only makes sense when using key path literals as a convenience to form closures; it quickly breaks down
+when using actual key path values as subscripts:
+
+```swift
+var mersennes = [0, 1, 3]
+primes[keyPath: \.append](mersennes, 7)  // Which is mutated: mersennes or primes?
+```
+
+This line of reasoning suggests that mutating methods are best left for a follow-up proposal:
+
+- If they are in fact limited to key path _literals_ converted to closures, then they are both outside the scope of this
+  proposal and not in conflict with it.
+- If we seek a broader solution, it would need to be consistent with unbound methods as well, and thus (as in the
+  previous section) this proposal is only uniformly respecting an existing problem, not introducing a new one.
+
+### Support for argument labels
+
+Accepting this proposal will create pressure to solve the argument label problem in the
+[wrapper type example above](#abstractions-over-properties-and-methods).
+
+One possibility would be loosening SE-0111 to make argument labels optional:
```

Comment:
I think you put it really well, nice job!

Reply:
Thanks!